react/src/component/util.js

```diff
@@ -0,0 +1,81 @@
+/* eslint eqeqeq: 0 */
+/* eslint no-unused-vars: 0 */
+import { useEffect, useLayoutEffect, useRef, useState } from 'react';
+
+export function constrainZoom(transformMatrix, size, dimension) {
+    if (transformMatrix[`scale${dimension}`] < 1) {
+        transformMatrix[`scale${dimension}`] = 1;
+        transformMatrix[`translate${dimension}`] = 0;
+    }
+    else {
+        if (transformMatrix[`translate${dimension}`] > 0) {
+            transformMatrix[`translate${dimension}`] = 0;
+        }
+        else if (size * transformMatrix[`scale${dimension}`] + transformMatrix[`translate${dimension}`] < size) {
+            transformMatrix[`translate${dimension}`] = size - size * transformMatrix[`scale${dimension}`];
+        }
+    }
+    return transformMatrix;
+}
+
+//TODO(pjm): use a library?
+export function debounce(fn, ms) {
+    let timer
+    return _ => {
+        clearTimeout(timer)
+        timer = setTimeout(_ => {
+            timer = null
+            fn.apply(this, arguments)
+        }, ms)
+    };
+}
+
+const xAxisSize = 30;
+const yAxisSize = 60;
+const margin = 25;
+
+function heightAdjustment() {
+    return xAxisSize + margin * 2;
+}
+
+function widthAdjustment() {
+    return yAxisSize + margin * 2;
+}
+
+export function graphMetrics(dim) {
+    return {
+        graphHeight: dim.height - heightAdjustment(),
+        graphWidth: dim.width - widthAdjustment(),
+        graphX: yAxisSize + margin,
+        graphY: margin,
+    };
+}
```

Comment:
```suggestion
export function graphContentBounds(dim) {
    return {
        height: dim.height - heightAdjustment(),
        width: dim.width - widthAdjustment(),
        offsetX: yAxisSize + margin,
        offsetY: margin,
    };
}
```

Let's use the word 'content' to differentiate between the size including the margin and axes and the size without those. I think we should avoid prefixes like 'graph' when we name fields within objects. This will help us unify types later. For instance, this is describing a 'rect' and anywhere else in Sirepo where we describe rectangles will likely have the same values.